system/HTTP/SiteURIFactory.php

```diff
@@ -0,0 +1,261 @@
+                break;
+        }
+
+        return ($routePath === '/' || $routePath === '') ? '/' : ltrim($routePath, '/');
+    }
+
+    /**
+     * Will parse the REQUEST_URI and automatically detect the URI from it,
+     * fixing the query string if necessary.
+     *
+     * This method updates superglobal $_SERVER and $_GET.
+     *
+     * @return string The route path (before normalization).
+     */
+    private function parseRequestURI(): string
+    {
+        if (! isset($this->server['REQUEST_URI'], $this->server['SCRIPT_NAME'])) {
+            return '';
+        }
+
+        // parse_url() returns false if no host is present, but the path or query
+        // string contains a colon followed by a number. So we attach a dummy
+        // host since REQUEST_URI does not include the host. This allows us to
+        // parse out the query string and path.
+        $parts = parse_url('http://dummy' . $this->server['REQUEST_URI']);
+        $query = $parts['query'] ?? '';
+        $path  = $parts['path'] ?? '';
+
+        // Strip the SCRIPT_NAME path from the URI
+        if (
+            $path !== '' && isset($this->server['SCRIPT_NAME'][0])
+            && pathinfo($this->server['SCRIPT_NAME'], PATHINFO_EXTENSION) === 'php'
+        ) {
+            // Compare each segment, dropping them until there is no match
+            $segments = $keep = explode('/', $path);
+
+            foreach (explode('/', $this->server['SCRIPT_NAME']) as $i => $segment) {
+                // If these segments are not the same then we're done
+                if (! isset($segments[$i]) || $segment !== $segments[$i]) {
+                    break;
+                }
+
+                array_shift($keep);
+            }
+
+            $path = implode('/', $keep);
+        }
+
+        // This section ensures that even on servers that require the URI to
+        // contain the query string (Nginx) a correct URI is found, and also
+        // fixes the QUERY_STRING Server var and $_GET array.
+        if (trim($path, '/') === '' && strncmp($query, '/', 1) === 0) {
+            $parts    = explode('?', $query, 2);
+            $path     = $parts[0];
+            $newQuery = $query[1] ?? '';
+
+            $this->server['QUERY_STRING'] = $newQuery;
+            $this->updateServer('QUERY_STRING', $newQuery);
+        } else {
+            $this->server['QUERY_STRING'] = $query;
+            $this->updateServer('QUERY_STRING', $query);
+        }
+
+        // Update our global GET for values likely to have been changed
+        parse_str($this->server['QUERY_STRING'], $get);
+        $this->updateGetArray($get);
+
+        return URI::removeDotSegments($path);
+    }
+
+    private function updateServer(string $key, string $value): void
+    {
+        $_SERVER[$key] = $value;
+    }
+
+    private function updateGetArray(array $array): void
+    {
+        $_GET = $array;
+    }
```

Comment:
These make me pretty uneasy. What about an external service dependency that handles Globals, so it can be mocked or replaced? This behavior within a Factory is very counterintuitive.

Reply:
What do you mean?
Adding another class that handles globals?
```php
    private $superglobals;

    private function updateServer(string $key, string $value): void
    {
        $this->superglobals->setServer($key, $value);
    }
```

Reply:
Yes, precisely. It could be adapted to anywhere else in the framework we touch globals.

Reply:
Added class `Superglobals`.